### PR TITLE
Gateway Manager Micorservice

### PR DESCRIPTION
**Pull request by songxiaoyan**
Current support CREATE,UPDATE,DELETE(for zeta gateway)，continued development based on the design will follow

### CODE REVIEW

**Review by cj-chung — CHANGES_REQUESTED** [review]
@songxiaoyan  I cannot find UT codes in this PR. Could you add UT in this PR to verify basic functionalities especially for DB (create/update/delete/query) operations?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    /**
+     * List All Available Gateways
+     *
+     * @param projectId
+     * @return
+     */
+    @GetMapping("/project/{projectid}/gateways")
+    public void getGatewaysByVpcId(@PathVariable("projectid") String projectId) {
+
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java:132]: @songxiaoyan The gateway here should be project-scope, not vpc-scope. So, it should be listed all available gateways in the current project.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GWAttachmentRepository.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Repository
+public class GWAttachmentRepository implements ICacheRepository<GWAttachment> {
+
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GWAttachmentRepository.java:22]: @songxiaoyan do we need cache for GWAttachment?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GatewayRepository.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Repository
+public class GatewayRepository implements ICacheRepository<GatewayEntity> {
+
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GatewayRepository.java:23]: @songxiaoyan Same here. Do we need cache for Gateway?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/RouteTableRepository.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.futurewei.alcor.gatewaymanager.dao;
+
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class RouteTableRepository {
+}
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/RouteTableRepository.java:7]: @songxiaoyan We probably don't need it.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/entity/GatewayWebJson.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.futurewei.alcor.web.entity.gateway.GatewayEntity;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+public class GatewayWebJson {
+
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/entity/GatewayWebJson.java:11]: If this class is the return for "list all available gateway", it's better rename it "GatewaysWebJson"

File: services/gateway_manager/src/main/resources/application.properties [hunk 6]
```diff
@@ -0,0 +1,22 @@
+server.port=9015
+
```
Comment (cj-chung) [inline services/gateway_manager/src/main/resources/application.properties:2]: @xieus Does this port for GM? please confirm.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+
+    public GatewayService() {
+    }
+
+    public GatewayInfo createGatewayInfo(String projectId, VpcInfo vpcInfo) throws Exception {
+
+        //TODO check whether the project type is zeta by projectId
+
+        // create GatewayInfo entity in the DB
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:47]: @songxiaoyan Please re-phrase the comment here. We are not creating a GatewayInfo enity in GM's DB.

File: services/gateway_manager/src/main/resources/application.properties (hunk 6 above)
Comment (xieus) [reply services/gateway_manager/src/main/resources/application.properties:2]: I am fine with 9015 and K8s cluster port of 30015. Keep in mind that 9014 remain available therefore we can keep it for other microservice.

File: web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayIpJson.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package com.futurewei.alcor.web.entity.gateway;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+public class GatewayIpJson {
+
```
Comment (cj-chung) [inline web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayIpJson.java:10]: Should we include "zeta" in the class name? since it's dedicated for zeta gateway only.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        executor.runAsync(restClinet::createVPCInZetaGateway, new VpcInfoSub(vpcInfo.getVpcId(), vpcInfo.getVpcVni()));
+        executor.runAsync((Supplier<Void>) () -> {
+            // wait result of all async
+            List<Object> result = executor.joinAllAsync();
+            // whether to rollback by checking result's size is 2
+            try {
+                updateDPMCacheGateway(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("update GatewayEntity status to READY failed, detail message: {}", e.getMessage());
+                //TODO how to handle this exception? rollback all?
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:65]: If the execution goes to this far, that's mean DPM already has an gatewayinfo entry for zeta with status = PENDING. In that case, when port creation comes to DPM, DPM should wait for the status become READY for 30s. If the status still PENDING or FAILED, DPM will switch to OVS plugin. So, should we let the updateDPMCacheGateway try some times and if still failed rollback GM's DB and ask Zeta to remove the corresponding gateway resources.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            }
+
+            try {
+                rollback(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("rollback failed, detail message: {}", e.getMessage());
+                //TODO if rollback failed,how we should handle it?
+            }
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:73]: @songxiaoyan If rollback failed, we should raise an alarm or error.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GWAttachmentRepository.java (hunk 2 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GWAttachmentRepository.java:22]: I think we need to create the Cache Cache of GatewayEntity and GwAttachment in GM's DB for later interface query

Comment (cj-chung) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GWAttachmentRepository.java:22]: Ok, agree.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/RouteTableRepository.java (hunk 4 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/RouteTableRepository.java:7]: ok, I can deleted reserved interface if don't need

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GatewayRepository.java (hunk 3 above)
Comment (cj-chung) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/dao/GatewayRepository.java:23]: I agree to have cache for GM's DB access.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/entity/GatewayWebJson.java (hunk 5 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/entity/GatewayWebJson.java:11]: ok

File: services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java [hunk 11]
```diff
@@ -35,10 +33,9 @@
 import com.futurewei.alcor.web.entity.ip.IpAddrRequest;
 import com.futurewei.alcor.web.entity.mac.MacStateJson;
 import com.futurewei.alcor.web.entity.port.PortEntity;
-import com.futurewei.alcor.web.entity.route.RouteEntity;
+import com.futurewei.alcor.web.entity.route.RouteWebJson;
 import com.futurewei.alcor.web.entity.subnet.*;
```
Comment (cj-chung) [inline services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java:37]: @songxiaoyan why your code will touch the import in the SubnetController.java?

File: services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/service/SubnetService.java [hunk 12]
```diff
@@ -2,18 +2,15 @@
...
 import com.futurewei.alcor.common.exception.*;
-
 import com.futurewei.alcor.subnet.exception.*;
+import com.futurewei.alcor.web.entity.ip.IpAddrRequest;
+import com.futurewei.alcor.web.entity.mac.MacStateJson;
 import com.futurewei.alcor.web.entity.port.PortEntity;
-import com.futurewei.alcor.web.entity.route.InternalRouterInfo;
+import com.futurewei.alcor.web.entity.route.RouteWebJson;
 import com.futurewei.alcor.web.entity.subnet.NewHostRoutes;
 import com.futurewei.alcor.web.entity.subnet.SubnetEntity;
```
Comment (cj-chung) [inline services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/service/SubnetService.java:11]: @songxiaoyan Your IDE seems re-arrange the import statements in the code.

File: web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+public class GatewayEntity extends CustomerResource {
+    private GatewayType type;
+    private String status; // (available)
+    private List<GatewayIp> ips;
+    private List<String> attachments;
+    private List<String> routetables;
+    private List<String> tags;
+    private String owner;
+    private Map<String, String> options;
+}
```
Comment (cj-chung) [inline web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java:20]: @songxiaoyan Do we need constructor here?

File: services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/service/SubnetService.java (hunk 12 above)
Comment (songxiaoyan) [reply services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/service/SubnetService.java:11]: Merge code and resolve conflicts, it automatically import, I tried to turn it off.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java (hunk 9 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:65]: if update DPM's cache status to READY is fail,we need retry some times(3 times?),if still failed,we should rollback all the request resource, is this mean?

Comment (cj-chung) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:65]: Yes. that's what I think.

File: web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java (hunk 13 above)
Comment (songxiaoyan) [reply web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java:20]: you mean it should be constructor in the GM entity package?

Comment (cj-chung) [reply web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java:20]: I mean such the code like this:
```
public GatewayEntity() {
    }

    public GatewayEntity(GatewayType type, String status, List<GatewayIp> ips, private List<String> attachments, List<String> routetables, ...) {
... 
}
```

**Comment by songxiaoyan** [comment]
> @songxiaoyan I cannot find UT codes in this PR. Could you add UT in this PR to verify basic functionalities especially for DB (create/update/delete/query) operations?

I'll make it up as soon as possible

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java (hunk 1 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java:132]: This is a reserved interface, which will be adjusted according to actual business in the future

File: web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java (hunk 13 above)
Comment (songxiaoyan) [reply web/src/main/java/com/futurewei/alcor/web/entity/gateway/GatewayEntity.java:20]: oh,I get your mean,I use the set method to assign values, if necessary, I can add the construction method

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    public void updateGatewayInfoForZeta(String projectId, GatewayInfo newGatewayInfo) throws Exception {
+        //TODO check whether the project type is zeta by projectId
+
+        GatewayEntity gatewayEntity = null;
+        Map<String, GWAttachment> attachmentsMap = gwAttachmentRepository.findAllItems();
+        for (GatewayEntity newGatewayEntity : newGatewayInfo.getGatewayEntities()) {
+            for (GWAttachment attachment : attachmentsMap.values()) {
+                if (attachment.getResourceId().equals(newGatewayInfo.getResourceId())) {
```
Comment (chenpiaoping) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:86]: Should be attachment.getGatewayId() == newGatewayInfo.getResourceId() ?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Update gatewayInfo (for zeta gateway)
+     *
+     * @param projectId
+     * @param vpcId
+     * @return
+     */
+    @PutMapping("/project/{projectid}/gatewayinfo/{resource_id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public ResponseId updateGatewayInfoForZeta(@PathVariable("projectid") String projectId, @PathVariable("resource_id") String vpcId, @RequestBody GatewayInfoJson gatewayInfoJson) throws Exception {
```
Comment (chenpiaoping) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java:48]: Who will call this interface? vpc-manager?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+                    gatewayEntity = gatewayRepository.findItem(attachment.getGatewayId());
+                    if (gatewayEntity == null) {
+                        throw new Exception(ExceptionMsgConfig.GATEWAY_ENTITY_NOT_FOUND.getMsg());
+                    }
+                    if (gatewayEntity.getType().equals(newGatewayEntity.getType())) {
+                        gatewayEntity.setStatus(newGatewayEntity.getStatus());
```
Comment (chenpiaoping) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:92]: need gatewayRepository.addItem(gatewayEntity) here ?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/AttachmentService.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            throw new Exception(ExceptionMsgConfig.ATTACHMENT_NOT_FOUND.getMsg());
+        }
+        GatewayEntity gatewayEntity = gatewayRepository.findItem(gatewayId);
+        if (gatewayEntity == null) {
+            throw new Exception(ExceptionMsgConfig.GATEWAY_ENTITY_NOT_FOUND.getMsg());
+        }
+        if (!gatewayId.equals(attachment.getGatewayId())) {
+            throw new Exception(ExceptionMsgConfig.GATEWAY_NOT_ASSOCIATED_ATTACHMENT.getMsg());
+        }
+        gatewayEntity.getAttachments().removeIf(attachId::equals);
```
[the remark this answers is not in the document]
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/AttachmentService.java:45]: have Updates in the code below

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/AttachmentService.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+        //delete attachment and update GatewayEntity in the DB
+        gwAttachmentRepository.deleteItem(attachId,gatewayEntity);
```
[the remark this answers is not in the document]
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/AttachmentService.java:48]: There are transactions in the method

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+     * @return
+     * @throws Exception
+     */
+    @DeleteMapping("/project/{projectid}/gatewayinfo/{resource_id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public ResponseId deleteGatewayInfoForZeta(@PathVariable("projectid") String projectId, @PathVariable("resource_id") String vpcId) throws Exception {
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectId);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(vpcId);
+        gatewayService.deleteGatewayInfoForZeta(projectId, vpcId);
+        log.info("GatewayInfo deleted success,the resource_id is: {}", vpcId);
```
[the remark this answers is not in the document]
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java:71]: I used the @Slf4j annotation,do don't need to initialize log.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java (hunk 14 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:86]: String parameters,I think use the equals to compare is better

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java (hunk 16 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:92]: Have updates in the code below

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java (hunk 15 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java:48]: Not sure, now it’s just providing this interface

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/AttachmentController.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+
+    /**
+     * Create an attachment
+     *
+     * @param projectId
+     * @param gatewayId
+     * @param gwAttachment
+     */
+    @PostMapping("/project/{projectid}/gateways/{gateway_id}/attachments")
```
[the remark this answers is not in the document]
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/AttachmentController.java:25]: ok

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java (hunk 14 above)
Comment (chenpiaoping) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:86]: attachment.getGatewayId(), not attachment.getResourceId() ?

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/AttachmentService.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+
+    public List<GWAttachment> getAllAttachments(String gatewayId) throws CacheException {
+        Map<String, GWAttachment> attachmentMap = gwAttachmentRepository.findAllItems();
```
[the remark this answers is not in the document]
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/AttachmentService.java:53]: ok

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+            return null;
+        });
+        return gatewayInfo;
+    }
+
+    public void updateGatewayInfoForZeta(String projectId, GatewayInfo newGatewayInfo) throws Exception {
+        //TODO check whether the project type is zeta by projectId
+
+        GatewayEntity gatewayEntity = null;
+        Map<String, GWAttachment> attachmentsMap = gwAttachmentRepository.findAllItems();
```
[the remark this answers is not in the document]
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:83]: Query all, have not parameters.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java (hunk 14 above)
Comment (songxiaoyan) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:86]: attachment.getGatewayId() is the gatewayId, attachment.getResourceId() and newGatewayInfo.getResourceId()  is the vpcId

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java (hunk 15 above)
Comment (cj-chung) [reply services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/controller/GatewayController.java:48]: DPM will call this interface if there is no gatewayinfo entry in the DPM's cache for a vpc.

File: services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+            try {
+                rollback(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("rollback failed, detail message: {}", e.getMessage());
+                // If rollback failed, we should raise an alarm or error.
+            }
+            return null;
+        });
+        return gatewayInfo;
+    }
```
Comment (cj-chung) [inline services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java:75]: We can leave rollback to another PR and figure out a better generic way for rollback.

**Review by cj-chung — APPROVED** [review]
The current code looks good for zeta gateway integration. The rollback part can leave for another PR after we figure out a better generic way to handle the rollback.
